pkg/api/v1beta3/types.go

```diff
@@ -934,7 +934,7 @@ type ServiceSpec struct {
 	Ports []ServicePort `json:"ports" description:"ports exposed by the service"`
 
 	// This service will route traffic to pods having labels matching this selector. If null, no endpoints will be automatically created. If empty, all pods will be selected.
-	Selector map[string]string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
+	Selector string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
```

Comment:
----- Original Message -----
> > @@ -934,7 +934,7 @@ type ServiceSpec struct {
> >  	Ports []ServicePort `json:"ports" description:"ports exposed by the
> >  	service"`
> >  
> >  	// This service will route traffic to pods having labels matching this
> >  	selector. If null, no endpoints will be automatically created. If empty,
> >  	all pods will be selected.
> > -	Selector map[string]string `json:"selector" description:"label keys and
> > values that must match in order to receive traffic for this service; if
> > empty, all pods are selected, if not specified, endpoints must be manually
> > specified"`
> > +	Selector string `json:"selector" description:"label keys and values that
> > must match in order to receive traffic for this service; if empty, all
> > pods are selected, if not specified, endpoints must be manually
> > specified"`
> 
> I don't follow the last bit here.  The idea of a magic label in the
> selector-map that was a full selector-expression is kind of smelly.  What's
> the downside of accepting either a map or a string?  A map is a logical AND
> of equality expressions.  If that is good enough for you, go ahead an use
> it.  The string (or some other type) is the escape valve for when that is
> not enough (and is the normalized form of all selectors).

Is "" really "magic"?  I'm just thinking that maps are actually something a client library can deal with today (for simple cases), whereas the full label selector syntax requires you to write a parser in every language you need to deal with labels in.  I'm not saying the map is better, just that it does at least handle the simple and complicated cases well enough to ease someone in (you can say "i don't understand your complex case").

`{"": "label notin (value1,value2)"}` -> `{"": "label notin (value1,value2)", "label": ""}`

Having two different types for the same value complicates lots of other clients and swagger (polymorphic between struct and primitive) - we have enough trouble with JSON parsing in Go that I wouldn't want to inflict that on clients.  IntString is two primitive types for which clear conversions exist.

I think polymorphic value is probably slightly better than requiring two fields ('selector' and 'selectorQuery'), but it's nasty to your clients to change value types in rare occasions.  People write using the map syntax and it works, and then someone in another client changes it and they blow up (because the client author didn't realize that we'd change the type of a field dynamically).  Sounds like a production bug waiting to happen...